From 1e952194f8c10d8fe30cd12514e087e7207d1fd3 Mon Sep 17 00:00:00 2001
From: Ding Qiang <b02550@freescale.com>
Date: Mon, 20 Jul 2009 10:50:23 +0800
Subject: [PATCH 072/280] ENGR00114316 IPUv3:  Resources for IPC should be
 correctly released.

Modify the strategy of resouces release,
Only the Primary layer will release semaphore and shm.

Signed-off-by: DingQiang <b02550@freescale.com>
(cherry picked from commit aef7aa832c3d60f20629dee2ff81b23409d1c946)
---
 screenlayer/ScreenLayer.c | 25 +++++++++++++++++--------
 1 file changed, 17 insertions(+), 8 deletions(-)

diff --git a/screenlayer/ScreenLayer.c b/screenlayer/ScreenLayer.c
index fced42c..1ed3751 100644
--- a/screenlayer/ScreenLayer.c
+++ b/screenlayer/ScreenLayer.c
@@ -579,11 +579,20 @@ SLRetCode DestoryScreenLayer(ScreenLayer *pSL)
 	pthread_counter--;
 	if(pthread_counter == 0)
 	{
-		munmap(vshmSLPriv, 3*sizeof(ScreenLayerPriv));
-		vshmSLPriv = NULL;
-		sem_post(semID);
-		sem_unlink(semName);
-		shm_unlink(shmName);
+		if(pSLPriv->isPrimary)
+		{
+			munmap(vshmSLPriv, 3*sizeof(ScreenLayerPriv));
+			vshmSLPriv = NULL;
+			sem_post(semID);
+			sem_close(semID);
+			shm_unlink(shmName);
+		}
+		else
+		{
+			munmap(vshmSLPriv, 3*sizeof(ScreenLayerPriv));
+			vshmSLPriv = NULL;
+			sem_post(semID);
+		}
 		goto done;
 	}
 
@@ -994,7 +1003,7 @@ void* GetPrimarySLHandle(char * pFbdev)
 
 	if(vshmSLPriv == NULL)
 		ret = PreScreenLayerIPC(pFbdev);
-	if(ret != E_RET_SUCCESS)
+	if(ret != E_RET_SUCCESS || vshmSLPriv == NULL)
 	{
 		dbg(DBG_ERR, "Prepared semaphore & shm failed !\n");
 		return (void*)0;
@@ -1003,13 +1012,13 @@ void* GetPrimarySLHandle(char * pFbdev)
 	sem_wait(semID);
 	if(vshmSLPriv->isPrimary)
 	{
-		dbg(DBG_DEBUG, "GetPrimarySLHandle end1!\n");
+		dbg(DBG_DEBUG, "GetPrimarySLHandle is OK!\n");
 		sem_post(semID);
 		return	(void *)1;
 	}
 	else
 	{
-		dbg(DBG_DEBUG, "GetPrimarySLHandle end2!\n");
+		dbg(DBG_ERR, "GetPrimarySLHandle Error!\n");
 		sem_post(semID);
 		return  (void *)0;
 	}
-- 
1.8.0

